### PR TITLE
feature: Add wolf rbac plugin

### PR DESCRIPTION
**Pull request by iGeeky**
### Summary

`wolf-rbac` is an authentication and authorization (rbac) plugin. 
The rbac feature is provided by [wolf](https://github.com/iGeeky/wolf). For more information about wolf, please refer to [wolf documentation](https://github.com/iGeeky/wolf).

### CODE REVIEW

File: lua/apisix/plugins/wolf-rbac.lua [hunk 1]
```diff
@@ -0,0 +1,380 @@
+
+local core     = require("apisix.core")
+local ck       = require("resty.cookie")
+local consumer = require("apisix.consumer")
+local json     = require("apisix.core.json")
+local ngx_re = require("ngx.re")
+local cjson = require("cjson")
```
Comment (membphis) [inline lua/apisix/plugins/wolf-rbac.lua]: Maybe `core.json` is enough?

File: lua/apisix/plugins/wolf-rbac.lua [hunk 2]
```diff
@@ -0,0 +1,380 @@
...
+
+    local httpc = http.new()
+    if timeout then
+        httpc:set_timeout(timeout)
+    end
+
+    local params = {method = method, headers = myheaders, body=body, ssl_verify=false}
+    local res, err = httpc:request_uri(uri, params)
+    if err then
+        core.log.error("FAIL REQUEST [ ",core.json.delay_encode({method=method, uri=uri, body=body, headers=myheaders}), " ] failed! res is nil, err:", err)
```
Comment (membphis) [inline lua/apisix/plugins/wolf-rbac.lua]: this line is too long.

File: lua/apisix/plugins/wolf-rbac.lua [hunk 3]
```diff
@@ -0,0 +1,380 @@
...
+
+function _M.check_schema(conf)
+    core.log.info("input conf: ", core.json.delay_encode(conf))
+
+    local ok, err = core.schema.check(schema, conf)
+    if not ok then
+        return false, err
+    end
+
+    if not conf.appid then
```
Comment (membphis) [inline lua/apisix/plugins/wolf-rbac.lua]: We can set the default value by schema.

Here is an example: https://github.com/apache/incubator-apisix/blob/e6804360d1712d456cfee31b9d6abb8b16cfffba/lua/apisix/schema_def.lua#L76

File: lua/apisix/plugins/wolf-rbac.lua [hunk 4]
```diff
@@ -0,0 +1,380 @@
...
+    if not conf.server then
+        conf.server = 'http://127.0.0.1:10080'
+    end
+
+    return true
+end
+
+
+local function fetch_rbac_token()
+    local args = ngx.req.get_uri_args()
```
Comment (membphis) [inline lua/apisix/plugins/wolf-rbac.lua]: we can use `ctx.var` to fetch it. eg: `ctx.var.arg_ rbac_token`, `ctx.var.http_ Authorization`, ``ctx.var.cookie_x-rbac-token`

File: lua/apisix/plugins/wolf-rbac.lua [hunk 5]
```diff
@@ -0,0 +1,380 @@
...
+    end
+
+    local headers = ngx.req.get_headers()
+    if headers.Authorization then
+        return headers.Authorization
+    end
+    if headers['x-rbac-token'] then
+        return headers['x-rbac-token']
+    end
+    local cookie, err = ck:new()
```
Comment (membphis) [inline lua/apisix/plugins/wolf-rbac.lua]: please add a blank line

File: lua/apisix/plugins/wolf-rbac.lua [hunk 6]
```diff
@@ -0,0 +1,380 @@
...
+            errmsg = 'check permission failed!' .. tostring(err)
+            break
+        else
+            core.log.info("check permission request:", url, ", status:", res.status, ",body:", core.json.delay_encode(res.body))
+            if res.status < 500 then
+                break
+            else
+                core.log.info("request [curl -v ", url, "] failed! status:", res.status)
+                if i < retry_max then
+                    ngx.sleep(100)
```
Comment (membphis) [inline lua/apisix/plugins/wolf-rbac.lua]: `100` sec to too long, need a shorter time

File: lua/apisix/plugins/wolf-rbac.lua [hunk 7]
```diff
@@ -0,0 +1,380 @@
...
+                end
+            end
+        end
+    end
+
+    if err then
```
Comment (membphis) [inline lua/apisix/plugins/wolf-rbac.lua]: this is a bad code style, please take a look at this new style:

```lua
if err then
    -- ...
end

if res.status ~= 200 and res.status ~= 401 then
    return {status=500, err='request to wolf-server failed, status:' .. tostring(res.status)}
end

-- ... other code

```

File: lua/apisix/plugins/wolf-rbac.lua [hunk 8]
```diff
@@ -0,0 +1,380 @@
...
+        return {status=500, err='request to wolf-server failed, status:' .. tostring(res.status)}
+    end
+end
+
+
+function _M.rewrite(conf, ctx)
+    local url = ctx.var.uri
+    local action = ctx.var.request_method
+    local clientIP = core.request.get_ip(ctx)
+    local permItem = {action=action, url = url, clientIP = clientIP}
```
Comment (membphis) [inline lua/apisix/plugins/wolf-rbac.lua]: `action=action,` to `action = action,`

File: lua/apisix/plugins/wolf-rbac.lua [hunk 9]
```diff
@@ -0,0 +1,380 @@
...
+
+	local rbac_token, err = fetch_rbac_token()
+	if rbac_token == nil then
+		core.log.info("no permission to access ", core.json.delay_encode(permItem), ", need login!")
```
Comment (membphis) [inline lua/apisix/plugins/wolf-rbac.lua]: bad indentation

File: lua/apisix/plugins/wolf-rbac.lua [hunk 10]
```diff
@@ -0,0 +1,380 @@
...
+        return 401, {message = "Missing rbac token in request"}
+    end
+
+    local tokenInfo =parse_rbac_token(rbac_token)
```
Comment (membphis) [inline lua/apisix/plugins/wolf-rbac.lua]: please add a space before `parse_rbac_token`

File: lua/apisix/plugins/wolf-rbac.lua [hunk 11]
```diff
@@ -0,0 +1,380 @@
...
+    core.log.info("consumer: ", core.json.delay_encode(consumer))
+    local server = consumer.auth_conf.server
+
+    local url = ctx.var.uri
+    local action = ctx.var.request_method
+    local clientIP = core.request.get_ip(ctx)
+    local permItem = {appid=appid, action=action, url = url, clientIP = clientIP, wolf_token=wolf_token}
+
+    local res = check_url_permission(server, appid, action, url, clientIP, wolf_token)
+	core.log.info(" check_url_permission(", core.json.delay_encode(permItem), ") res: ",core.json.delay_encode(res))
```
Comment (membphis) [inline lua/apisix/plugins/wolf-rbac.lua]: Replace tabs with four spaces.

File: lua/apisix/plugins/wolf-rbac.lua [hunk 12]
```diff
@@ -0,0 +1,380 @@
...
+        local userInfo = res.userInfo
+        core.response.set_header("X-UserId", userInfo.id)
+        core.response.set_header("X-Username", userInfo.username)
+        core.response.set_header("X-Nickname", ngx.escape_uri(userInfo.nickname) or userInfo.username)
+        ctx.userInfo = userInfo
+		username = userInfo.username
+        nickname = userInfo.nickname
+	end
+
+	if res.status == 200 then
```
Comment (membphis) [inline lua/apisix/plugins/wolf-rbac.lua]: I think we should check `res.status` first.

File: lua/apisix/plugins/wolf-rbac.lua [hunk 13]
```diff
@@ -0,0 +1,380 @@
...
+        return core.response.exit(500, {message = "request to wolf-server failed!"})
+    end
+    if not body.ok then
+        core.log.error("user login [", request_debug, "] failed! response body:", core.json.delay_encode(body))
+        return core.response.exit(200, {message = body.reason})
+    end
+    core.log.info("user login [", request_debug, "] success! response body:", core.json.delay_encode(body))
+
+    local userInfo = body.data.userInfo
+    local wolf_token = body.data.token;
```
Comment (membphis) [inline lua/apisix/plugins/wolf-rbac.lua]: `;` is uesless

Comment (iGeeky) [reply lua/apisix/plugins/wolf-rbac.lua]: Done

File: lua/apisix/plugins/wolf-rbac.lua (hunk 12 above)
Comment (iGeeky) [reply lua/apisix/plugins/wolf-rbac.lua]: Done

Comment (iGeeky) [reply lua/apisix/plugins/wolf-rbac.lua]: res.status check is not necessary, no matter if res.status is 200 or 401,400, res.userInfo may return value

File: lua/apisix/plugins/wolf-rbac.lua (hunk 11 above)
Comment (iGeeky) [reply lua/apisix/plugins/wolf-rbac.lua]: Done

File: lua/apisix/plugins/wolf-rbac.lua (hunk 10 above)
Comment (iGeeky) [reply lua/apisix/plugins/wolf-rbac.lua]: Done

File: lua/apisix/plugins/wolf-rbac.lua (hunk 9 above)
Comment (iGeeky) [reply lua/apisix/plugins/wolf-rbac.lua]: Done

File: lua/apisix/plugins/wolf-rbac.lua (hunk 4 above)
Comment (iGeeky) [reply lua/apisix/plugins/wolf-rbac.lua]: Done

File: lua/apisix/plugins/wolf-rbac.lua (hunk 3 above)
Comment (iGeeky) [reply lua/apisix/plugins/wolf-rbac.lua]: Done

File: lua/apisix/plugins/wolf-rbac.lua (hunk 2 above)
Comment (iGeeky) [reply lua/apisix/plugins/wolf-rbac.lua]: Done

File: lua/apisix/plugins/wolf-rbac.lua (hunk 1 above)
Comment (iGeeky) [reply lua/apisix/plugins/wolf-rbac.lua]: Done

File: doc/plugins/wolf-rbac-cn.md [hunk 14]
```diff
@@ -0,0 +1,207 @@
...
+- [**属性**](#属性)
+- [**依赖项**](#依赖项)
+- [**如何启用**](#如何启用)
+- [**测试插件**](#测试插件)
+- [**禁用插件**](#禁用插件)
+
+## 名字
+
+`wolf-rbac`是一个认证及授权(rbac)插件，它需要与 `consumer` 一起配合才能工作。同时需要添加 `wolf-rbac` 到一个 `service` 或 `route`中。
+rbac功能由[wolf](https://github.com/iGeeky/wolf)提供, 有关wolf的更多信息, 请参考[wolf文档](https://github.com/iGeeky/wolf)。
```
[the remark this answers is not in the document]
Comment (iGeeky) [reply doc/plugins/wolf-rbac-cn.md]: Done

File: lua/apisix/plugins/wolf-rbac.lua [hunk 15]
```diff
@@ -0,0 +1,376 @@
...
+
+    if ctx.var.http_x_rbac_token then
+        return ctx.var.http_x_rbac_token
+    end
+
+    return ctx.var['cookie_x-rbac-token']
+end
+
+local function loadjson(str)
+    local ok, jso = pcall(function() return json.decode(str) end)
```
Comment (membphis) [inline lua/apisix/plugins/wolf-rbac.lua]: please pay attention to `json` and `json.safe`, it is a little different. https://github.com/apache/incubator-apisix/blob/master/lua/apisix/core/json.lua#L28

`local data, err = require("cjson.safe").decode(str)`
so we do not need a `pcall` here.

File: lua/apisix/plugins/wolf-rbac.lua [hunk 16]
```diff
@@ -0,0 +1,376 @@
...
+    if ok then
+        return jso
+    else
+        return nil, jso
+    end
+end
+
+local function check_url_permission(server, appid, action, resName, clientIP, wolf_token)
+    local retry_max = 3
+    local errmsg = nil
```
Comment (membphis) [inline lua/apisix/plugins/wolf-rbac.lua]: `local errmsg = nil` to `local errmsg`

new code is simpler

File: lua/apisix/plugins/wolf-rbac.lua [hunk 17]
```diff
@@ -0,0 +1,376 @@
...
+        return 401, {message = "Missing rbac token in request"}
+    end
+
+    local tokenInfo = parse_rbac_token(rbac_token)
+    core.log.info("token info: ", core.json.delay_encode(tokenInfo))
+    if tokenInfo.err then
+        return 401, {message = 'invalid rbac token: parse failed'}
+    end
+
+
```
Comment (membphis) [inline lua/apisix/plugins/wolf-rbac.lua]: one blank line here

File: lua/apisix/plugins/wolf-rbac.lua [hunk 18]
```diff
@@ -0,0 +1,376 @@
...
+    end
+
+    if res.status == 200 then
+        ---
+    else
+        -- no permission.
+        core.log.error(" check_url_permission(", core.json.delay_encode(permItem), ") failed, res: ",core.json.delay_encode(res))
+        return 401, {message = res.err, username = username, nickname = nickname}
+    end
+    core.log.info("hit wolf-rbac rewrite")
```
Comment (membphis) [inline lua/apisix/plugins/wolf-rbac.lua]: move this line before https://github.com/apache/incubator-apisix/pull/1095/files#diff-0d1c836712de65dc51f6498601f35938R216 is better.

File: lua/apisix/plugins/wolf-rbac.lua [hunk 19]
```diff
@@ -0,0 +1,376 @@
...
+        ctx.userInfo = userInfo
+        username = userInfo.username
+        nickname = userInfo.nickname
+    end
+
+    if res.status == 200 then
+        ---
+    else
+        -- no permission.
+        core.log.error(" check_url_permission(", core.json.delay_encode(permItem), ") failed, res: ",core.json.delay_encode(res))
```
Comment (membphis) [inline lua/apisix/plugins/wolf-rbac.lua]: this line is too long, need a shorter style

File: lua/apisix/plugins/wolf-rbac.lua [hunk 20]
```diff
@@ -0,0 +1,376 @@
...
+        return 401, {message = res.err, username = username, nickname = nickname}
+    end
+    core.log.info("hit wolf-rbac rewrite")
+end
+
+local function get_args()
+    local args, err
+    ngx.req.read_body()
+    if string.find(ngx.req.get_headers()["Content-Type"] or "","application/json", 0) then
```
Comment (membphis) [inline lua/apisix/plugins/wolf-rbac.lua]: We can turn off the pattern matching feature by using the optional fourth argument plain.

`> = string.find("Hello Lua user", "%su", 1, true) -- turn on plain searches, now not found`

http://lua-users.org/wiki/StringLibraryTutorial

Comment (membphis) [reply lua/apisix/plugins/wolf-rbac.lua]: `ctx.var.http_content_type` you can make a try in this way which is faster.

File: lua/apisix/plugins/wolf-rbac.lua [hunk 21]
```diff
@@ -0,0 +1,376 @@
...
+        core.log.error("login request [", request_debug, "] failed! err: ", err)
+        return core.response.exit(500, {message = "request to wolf-server failed! " .. tostring(err)})
+    end
+    core.log.info("login request [", request_debug, "] status: ", res.status, ", body: ", res.body)
+
+    if res.status ~= 200 then
+        core.log.error("login request [", request_debug, "] failed! status: ", res.status)
+        return core.response.exit(500, {message = "request to wolf-server failed! status:" .. tostring(res.status) })
+    end
+    local body = json.decode(res.body)
```
Comment (membphis) [inline lua/apisix/plugins/wolf-rbac.lua]: `local body, err = ...`

File: lua/apisix/plugins/wolf-rbac.lua (hunk 20 above)
Comment (iGeeky) [reply lua/apisix/plugins/wolf-rbac.lua]: how to get the `ctx` in a api function?

File: lua/apisix/plugins/wolf-rbac.lua (hunk 21 above)
Comment (iGeeky) [reply lua/apisix/plugins/wolf-rbac.lua]: Done

File: lua/apisix/plugins/wolf-rbac.lua (hunk 15 above)
Comment (iGeeky) [reply lua/apisix/plugins/wolf-rbac.lua]: Done

File: lua/apisix/plugins/wolf-rbac.lua (hunk 16 above)
Comment (iGeeky) [reply lua/apisix/plugins/wolf-rbac.lua]: Done

File: lua/apisix/plugins/wolf-rbac.lua (hunk 17 above)
Comment (iGeeky) [reply lua/apisix/plugins/wolf-rbac.lua]: Done

File: lua/apisix/plugins/wolf-rbac.lua (hunk 18 above)
Comment (iGeeky) [reply lua/apisix/plugins/wolf-rbac.lua]: Done

File: lua/apisix/plugins/wolf-rbac.lua (hunk 19 above)
Comment (iGeeky) [reply lua/apisix/plugins/wolf-rbac.lua]: Done

File: lua/apisix/plugins/wolf-rbac.lua (hunk 7 above)
Comment (iGeeky) [reply lua/apisix/plugins/wolf-rbac.lua]: Done

File: lua/apisix/plugins/wolf-rbac.lua [hunk 22]
```diff
@@ -0,0 +1,386 @@
...
+    local permItem = {action = action, url = url, clientIP = clientIP}
+    core.log.info("hit wolf-rbac rewrite")
+
+    local rbac_token = fetch_rbac_token(ctx)
+    if rbac_token == nil then
+        core.log.info("no permission to access ", core.json.delay_encode(permItem), ", need login!")
+        return 401, {message = "Missing rbac token in request"}
+    end
+
+    local tokenInfo = parse_rbac_token(rbac_token)
```
Comment (membphis) [inline lua/apisix/plugins/wolf-rbac.lua]: `local token, err = parse_rbac_token(rbac_token)`

this style is better

File: lua/apisix/plugins/wolf-rbac.lua [hunk 23]
```diff
@@ -0,0 +1,386 @@
...
+    core.log.info("token info: ", core.json.delay_encode(tokenInfo))
+    if tokenInfo.err then
+        return 401, {message = 'invalid rbac token: parse failed'}
```
Comment (membphis) [inline lua/apisix/plugins/wolf-rbac.lua]: `{message = 'invalid rbac token: parse failed, ' .. err}`

Should we return the error message?

File: lua/apisix/plugins/wolf-rbac.lua [hunk 24]
```diff
@@ -0,0 +1,386 @@
...
+        core.response.set_header("X-Nickname", ngx.escape_uri(userInfo.nickname) or userInfo.username)
+        ctx.userInfo = userInfo
+        username = userInfo.username
+        nickname = userInfo.nickname
+    end
+
+    if res.status ~= 200 then
+        -- no permission.
+        core.log.error(" check_url_permission(", core.json.delay_encode(permItem),
+            ") failed, res: ",core.json.delay_encode(res))
```
Comment (membphis) [inline lua/apisix/plugins/wolf-rbac.lua]: bad indentation

File: lua/apisix/plugins/wolf-rbac.lua [hunk 25]
```diff
@@ -0,0 +1,386 @@
...
+        return core.response.exit(400, {message = "appid [" .. tostring(appid) .. "] not found"})
+    end
+
+    core.log.info("consumer: ", core.json.delay_encode(consumer))
+
+    local uri = consumer.auth_conf.server .. '/wolf/rbac/login.rest'
+    local headers = new_headers()
+    headers["Content-Type"] = "application/json; charset=utf-8"
+    local timeout = 1000 * 5
+    local request_debug = core.json.delay_encode(
```
Comment (membphis) [inline lua/apisix/plugins/wolf-rbac.lua]:
```lua
local request_debug = core.json.delay_encode(
        {method = 'POST', uri = uri, body = args, headers = headers,timeout = timeout}
)
```

this style is clearer

File: lua/apisix/plugins/wolf-rbac.lua (hunk 22 above)
Comment (iGeeky) [reply lua/apisix/plugins/wolf-rbac.lua]: Done

Comment (iGeeky) [reply lua/apisix/plugins/wolf-rbac.lua]: Done

File: lua/apisix/plugins/wolf-rbac.lua (hunk 23 above)
Comment (iGeeky) [reply lua/apisix/plugins/wolf-rbac.lua]: Done

File: lua/apisix/plugins/wolf-rbac.lua (hunk 24 above)
Comment (iGeeky) [reply lua/apisix/plugins/wolf-rbac.lua]: Done

File: lua/apisix/plugins/wolf-rbac.lua (hunk 25 above)
Comment (iGeeky) [reply lua/apisix/plugins/wolf-rbac.lua]: Done

**Comment by membphis** [comment]
@iGeeky many thx

**Comment by moonming** [comment]
@iGeeky please add this plugin to `README.md` and `doc/README.md`, in order to help to find it.

**Comment by iGeeky** [comment]
@moonming @membphis thx. 
I created a new pr to add it.

**Pull request merged**